### ISSUE TITLE
Heap allocation in Box optimized out with rustc 1.60 but not with 1.61+

### ISSUE THREAD

**Issue by Nekrolm** [rust-lang/rust#98679]
I tried this code:

```rust
pub struct BBox {
    top: i32,
    left: i32,
    height: i32,
    width: i32
}

fn fiddle(bbox: BBox) -> Box<BBox> {
    Box::new(BBox {
        top: bbox.top * 2,
        ..bbox
    })
}

pub fn main() {
    let bbox = Box::new(BBox {
            top: 551,
            left: 100,
            height: 200,
            width: 180,
    });
    let bbox = fiddle(*bbox);
    let bbox_top = bbox.top;
    println!("{}", bbox_top)
}
```

With rustc 1.60 all Box allocations are totally optimized out.
But starting from 1.61 it doesn't happen.

[godbolt link](https://godbolt.org/z/nvob7acWW)

Possible it is the same issue as https://github.com/rust-lang/rust/issues/97751

**Comment by bugadani** [comment]
Interestingly, a much simpler case isn't optimized at all with either compiler version (though generated code changed between the versions): https://godbolt.org/z/5G9cP5n88

**Comment by erikdesjardins** [comment]
@bugadani that happens because println takes references to its arguments, which keeps the allocation alive. If you force a move then it is optimized out: https://godbolt.org/z/qz8n7M5jd

**Comment by erikdesjardins** [comment]
Since this starts getting optimized in 1.59 and stops in 1.61, it is likely #92419, which landed in 1.59 and was reverted in 1.61. (see issues linked in #94823 for more info)

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2398679.20Heap.20allocation.20in.20Box.20optimized.20out.20with.20rustc.201.2E60.E2.80.A6)).

@rustbot label -I-prioritize +P-medium

**Comment by erikdesjardins** [comment]
This was fixed in 1.75 by #102099: https://godbolt.org/z/zqcqh3Mch

**Comment by apiraino** [comment]
Given the age of this issue, I'm going to close it and assume it's fixed. @Nekrolm please feel free to open a new issue if it's not the case.

Closed by #102099
